### PR TITLE
feat(dashboard): smart project picker for the wizard ProjectDir step (closes #127)

### LINKED ISSUE #127: [RFC] Smart Project Picker for the wizard's Project dir step
## Labels (suggested)
`enhancement`, `dashboard`, `ux`, `rfc`

## Problem

The wizard's Project dir step (`main.rs:1063+`) is the only step that uses a free-text input — every other step is a navigable list (`j/k` + Enter). Three pain points compound:

1. The field is empty on first use, so the user has to type a full absolute path (35-50 chars on macOS).
2. Zellij plugin WASM hosts deliver keystrokes as `BareKey::Char(c)` one at a time, so `Cmd+V` paste is dropped. The user has to type each character.
3. Tab completion has a relative-path bug (#125).

## Proposed UX

Replace the text input with a navigable list seeded from recently-used `project_dir` values (extracted from the persisted state), with a free-text fallback for the long tail. Consistent with the other wizard steps.

```
┌─ Step 6/7: Project dir ───────────────────────────────┐
│ Filter:                                               │
│                                                       │
│ RECENTS                                               │
│ > /Users/me/Projects/Personal/synoptic    (just now)  │
│   /Users/me/Projects/Personal/hermes      (2h ago)    │
│   /Users/me/Work/cliente-x-api            (yesterday) │
│                                                       │
│ [type]: filter   [Enter]: select   [i]: free text     │
└───────────────────────────────────────────────────────┘
```

Typing filters in place (substring + subsequence, case-insensitive). `i` switches to free-text input (current behavior preserved as escape hatch).

## Design

**State.** Extend `WizardState` (`types.rs:131`) with `available_project_dirs: Vec<RecentProject>`, `project_dir_index: usize`, `project_dir_filter: String`, `project_dir_mode: { List, Input }`. Mirrors the existing AgentType/SandboxBackend/SandboxLevel/Provider pattern.

**Storage.** Extend `SavedState` (`state_file.rs`, schema v3→v4) with `global_recent_project_dirs: Vec<RecentProject>` and a `record_project_use(path)` helper called from `spawn_agent_custom`. Backward-compatible via `#[serde(default)]`, same pattern v1→v2→v3 used.

**Filter.** In-house ~40 LOC: case-insensitive substring + subsequence with a small score boost for path-segment-start matches. No new crate — consistent with the current minimal Cargo.toml (`zellij-tile`, `serde`, `serde_json`, `toml`).

**Skip rules.** `active_steps()` (`types.rs:223`) unchanged: `ProjectDir` stays as the always-present step, just with a different sub-UI.

## Out of scope (v1)

- **Git auto-discovery** (`find $HOME -name .git`): adds latency, ignore-list maintenance, iCloud/symlink edge cases. Worth its own RFC if v1 lands.
- **Declarative bookmarks** (`[[dashboard.projects]]` with per-project default agent/sandbox): orthogonal, design-heavy.
- **Fuzzy ranking on par with fzf**: in-house scorer is enough for tens of entries.

## Backward compat

A new config key gates the picker, default `true`:

```toml
[dashboard]
project_picker = true
```

`project_picker = false` keeps the legacy text input. Existing `default_project_dir` keeps working as the initial value for free-text mode.

## Open questions

1. **Recents storage**: per-launch-dir `.lince-dashboard` (where state already lives, but recents wouldn't survive launching from a different dir), or a new global `~/.config/lince-dashboard/recents.json`?
2. **Default**: should the picker be on by default for new installs, or opt-in?
3. **One step or two?**: my proposal transforms the existing `ProjectDir` step into a dual-mode picker. A separate `ProjectPicker` step before `ProjectDir` would keep the old behavior unchanged but adds a step in `active_steps()`.

## Related

- #125 — completer relative-path bug; prerequisite for the picker's free-text fallback.
- #126 — wizard input validation; prerequisite for the same reason.

### PR DESCRIPTION
**Pull request by rinaldofesta**
## Problem

The wizard's Project dir step was the only free-text step: empty on first use, no paste in the Zellij WASM host, so the user had to type a full absolute path every time.

## Fix

A dual-mode ProjectDir step:

- **List mode** — a navigable, filter-as-you-type list of recent project dirs (most-recently-used first). Type to filter (substring / subsequence, case-insensitive), `↑/↓` to move, `Enter` to select.
- **Input mode** — the legacy free-text entry with Tab completion and absolute-path validation, reached via `i`, or automatically when there are no recents yet.

`i` switches to free-text **only as the first keystroke** (filter still empty), so paths containing "i" (most of them) stay filterable once you start typing.

The focused/selected agent's workdir is moved to the front of the recents so it's pre-highlighted (context default, cf. #168).

## Storage

Recents live in a single **global** file — `~/.config/lince-dashboard/recents.json` (MRU order, capped at 20) — so they follow the user regardless of which directory the dashboard was launched from. Loaded async at startup; recorded on **user-initiated spawns only** (never on session restore).

Timestamps are intentionally omitted: wall-clock is unreliable under WASI, and MRU order is all the picker needs. (Relative-time display can be a follow-up.)

## Per your feedback on #127

> Let's avoid the backward compatibility option; it just messes the config, and we have the "i" fallback to the old behaviour. And consider default based on the selected agent as per #168.

- **No config gate** — the picker is always on; `i` is the fallback to the legacy input.
- **Context default** — focused agent's workdir pre-selected (#168 direction).

## Files

`recents.rs` (new, +unit tests for the MRU/parse logic), `types.rs`, `main.rs`, `dashboard.rs`. Builds clean for `wasm32-wasip1`.

## Out of scope (per the RFC)

Git auto-discovery, declarative bookmarks, fzf-grade ranking — each its own follow-up if this lands.

Closes #127

### CODE REVIEW

File: lince-dashboard/plugin/src/types.rs [hunk 1]
```diff
@@ -156,6 +156,31 @@ pub enum WizardStep {
...
+    /// Free-text path input (legacy behavior; reached via `i`, or used
+    /// automatically when there are no recents yet).
+    Input,
+}
+
+/// Case-insensitive match used by the project-dir picker filter: a hit is
+/// either a substring match or an in-order subsequence match (so `snptc`
+/// matches `synoptic`). Both `haystack` and `needle` must be lowercased by
+/// the caller.
+pub fn project_dir_filter_matches(haystack: &str, needle: &str) -> bool {
```
Comment (maeste) [inline lince-dashboard/plugin/src/types.rs:175]: 💡 **Suggestion:** The subsequence matching logic here is concise and correct. A one-line doc-example (e.g. `// "sntp" matches "synoptic"`) directly in the fn body or as a doc attribute would help future maintainers instantly recognise this as fuzzy/subsequence matching rather than a bug.

File: lince-dashboard/plugin/src/recents.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+/// Parse the recents file contents. Tolerant of a missing/empty/garbage file:
+/// any parse failure yields an empty list rather than an error (recents are a
+/// convenience, never load-blocking).
+pub fn parse_loaded_recents(stdout: &[u8]) -> Vec<String> {
+    let content = String::from_utf8_lossy(stdout);
+    let trimmed = content.trim();
+    if trimmed.is_empty() {
+        return Vec::new();
+    }
+    serde_json::from_str::<Vec<String>>(trimmed).unwrap_or_default()
```
Comment (maeste) [inline lince-dashboard/plugin/src/recents.rs]: 💡 **Suggestion:** Since this function is documented as tolerant of malformed input, adding `.dedup_by(|a, b| a == b)` before returning would defend against duplicate entries if the file is manually edited or corrupted outside of the normal push_recent → save_recents_async path. Purely defensive — push_recent already deduplicates on write.

File: lince-dashboard/plugin/src/dashboard.rs [hunk 3]
```diff
@@ -965,64 +965,122 @@ pub fn render_wizard(
...
+                    let max_show = 6;
+                    let total = filtered.len();
+                    let sel = wizard.project_dir_index.min(total.saturating_sub(1));
+                    let start = if sel >= max_show { sel + 1 - max_show } else { 0 };
+                    for (i, entry) in filtered.iter().enumerate().skip(start).take(max_show) {
+                        let marker = if i == sel { ">" } else { " " };
+                        let short = collapse_tilde(entry);
+                        // Truncate from the left to keep the distinguishing tail.
+                        let max_len = box_width.saturating_sub(6);
+                        let disp = if short.chars().count() > max_len && max_len > 1 {
```
Comment (maeste) [inline lince-dashboard/plugin/src/dashboard.rs]: 💡 **Suggestion:** This left-truncation-with-UTF-8-boundary block is duplicated verbatim at line 1046 (completion renderer) and also existed in the pre-PR code. Consider extracting a small `fn truncate_left(s: &str, max_len: usize) -> String` helper to keep the rendering logic DRY.

**Review by maeste — APPROVED** [review]
## Code Review Summary

**Verdict: Approve** — Clean implementation, well-scoped, good test coverage on the new module.

### ✅ Looks Good

- **Dual-mode design** (List + Input) is elegant. The `i` gate (mode-switch only on empty filter) is a smart UX choice that avoids stealing the letter `i` from normal filtering.
- **Context default** (#168 direction): biasing the focused agent's workdir to the front of the recents list feels natural.
- **Graceful degradation**: no recents yet → straight to Input mode (legacy intact); filter matches nothing → hand typed text to Input mode instead of dead end; Backspace on empty Input field → return to List mode when available.
- **`recents.rs` is solid**: tolerant parser (`unwrap_or_default`), capped MRU list, trailing-slash normalization, empty/root rejection, fire-and-forget async save with heredoc (correct pattern for WASI sandbox).
- **Unit tests** cover all edge cases of `push_recent` and `parse_loaded_recents`: missing file, garbage JSON, round-trip, dedup, cap-at-max, empty rejection.
- **Subsequence filter** (`snptc` matching `synoptic`) is a nice power-user touch without added complexity.
- **No security concerns**: paths come from the agent's own `project_dir` field (set by the user or the system), not from untrusted input. Heredoc delimiter `__LINCE_EOF__` is single-quoted so no expansion risk.

### 💡 Suggestions (non-blocking)

- **`types.rs:173`** — `project_dir_filter_matches`: consider adding a doc-example comment showing the subsequence behaviour. The doc comment above already mentions it, but an assert-style example would make it immediately obvious to future readers.
- **`recents.rs:58`** — `parse_loaded_recents`: a defensive `.dedup_by()` after parsing would protect against corrupted files introducing duplicates in the picker. Belt-and-suspenders since `push_recent` already deduplicates on write.
- **`dashboard.rs:997 vs 1046`** — left-truncation with UTF-8 boundary awareness is duplicated between the recents-list renderer and the completion renderer. Extracting a `truncate_left()` helper would eliminate the copy-paste.

**Comment by maeste** [comment]
LGTM, please rebase and solve conflicts and I'll merge it

**Comment by rinaldofesta** [comment]
Rebased onto latest `main` and resolved conflicts — ready to merge.

The interesting part was integrating with #169 (emoji markers + dir-based naming #167 + ProjectDir prefill #168), which landed on `main` and touched the same wizard code. Rather than pick one, I merged both:

- `WizardState` keeps **both** `project_dir_suggested` (#168) and the picker fields (#127).
- The ProjectDir step is dual-mode: **List** (recents picker) + **Input** (legacy free-text). The Input mode preserves #168's "first Backspace clears the pristine pre-fill" and #167's dir-derived default-name refresh; the List mode does the same name refresh on selection. Empty-field Backspace in Input mode now backs out to the recents list.

Also applied your three review suggestions:
1. **types.rs** — inline comment documenting the subsequence match (`"sntp" → "synoptic"`).
2. **recents.rs** — defensive, order-preserving de-dup in `parse_loaded_recents` (+ a test) for hand-edited/corrupted files.
3. **dashboard.rs** — extracted a `truncate_left()` helper, now shared by the completion renderer and the recents list (DRY).

Verified: full clean rebuild for `wasm32-wasip1`, zero warnings. Over to you for the merge 🙏
